aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SigV4aAuthDecorator.kt

```diff
@@ -0,0 +1,147 @@
+private class V4aAuthServiceRuntimePluginCustomization(private val codegenContext: ClientCodegenContext) :
+    ServiceRuntimePluginCustomization() {
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val codegenScope by lazy {
+        val awsRuntime = AwsRuntimeType.awsRuntime(runtimeConfig)
+        arrayOf(
+            "SigV4aAuthScheme" to awsRuntime.resolve("auth::sigv4a::SigV4aAuthScheme"),
+            "SharedAuthScheme" to RuntimeType.smithyRuntimeApi(runtimeConfig).resolve("client::auth::SharedAuthScheme"),
+        )
+    }
+
+    override fun section(section: ServiceRuntimePluginSection): Writable = writable {
+        when (section) {
+            is ServiceRuntimePluginSection.RegisterRuntimeComponents -> {
+                val serviceHasEventStream = codegenContext.serviceShape.hasEventStreamOperations(codegenContext.model)
+                if (serviceHasEventStream) {
+                    // enable the aws-runtime `sign-eventstream` feature
+                    addDependency(AwsCargoDependency.awsRuntime(runtimeConfig).withFeature("event-stream").toType().toSymbol())
+                }
+                section.registerAuthScheme(this) {
+                    rustTemplate("#{SharedAuthScheme}::new(#{SigV4aAuthScheme}::new())", *codegenScope)
+                }
+            }
+
+            else -> {}
+        }
+    }
+}
+
+private class V4aAuthOperationCustomization(private val codegenContext: ClientCodegenContext) : OperationCustomization() {
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val codegenScope by lazy {
+        val awsRuntime = AwsRuntimeType.awsRuntime(runtimeConfig)
+        arrayOf(
+            "SigV4aOperationSigningConfig" to awsRuntime.resolve("auth::sigv4a::SigV4aOperationSigningConfig"),
+            "SigningOptions" to awsRuntime.resolve("auth::sigv4a::SigningOptions"),
+            "SignableBody" to AwsRuntimeType.awsSigv4(runtimeConfig).resolve("http_request::SignableBody"),
+        )
+    }
+    private val serviceIndex = ServiceIndex.of(codegenContext.model)
+
+    override fun section(section: OperationSection): Writable = writable {
+        when (section) {
+            is OperationSection.AdditionalRuntimePluginConfig -> {
+                val authSchemes = serviceIndex.getEffectiveAuthSchemes(codegenContext.serviceShape, section.operationShape)
+                if (authSchemes.containsKey(SigV4Trait.ID)) {
+                    val unsignedPayload = section.operationShape.hasTrait<UnsignedPayloadTrait>()
+                    val doubleUriEncode = unsignedPayload || !disableDoubleEncode(codegenContext.serviceShape)
+                    val contentSha256Header = needsAmzSha256(codegenContext.serviceShape)
+                    val normalizeUrlPath = !disableUriPathNormalization(codegenContext.serviceShape)
+                    rustTemplate(
+                        """
+                        let mut signing_options = #{SigningOptions}::default();
```

Comment:
I'm thinking we should combine the V4 and V4a decorators so that this signing option initialization only needs to happen once (assuming the supporting types are merged).